src/service.rs

```diff
@@ -424,14 +832,312 @@ impl Service {
                         self.send_ping(enr);
                     }
                 }
+                Some(Ok((active_topic, active_ticket))) = self.tickets.next() => {
+                    let enr = self.local_enr.read().clone();
+                    // When the ticket time expires a new REGTOPIC request is automatically sent to the
+                    // ticket issuer and the registration attempt stays in the [`RegistrationState::Ticket`]
+                    // from sending the first REGTOPIC request to this contact for this topic.
+                    self.reg_topic_request(active_ticket.contact(), active_topic.topic().clone(), enr, Some(active_ticket.ticket()));
+                }
+                Some(topic_query_progress) = self.active_topic_queries.next() => {
+                    match topic_query_progress {
+                        TopicQueryState::Finished(topic_hash) | TopicQueryState::TimedOut(topic_hash) => {
+                            if let Some(query) = self.active_topic_queries.queries.remove(&topic_hash) {
+                                if let Some(callback) = query.callback {
+                                    if callback.send(Ok(query.results.into_values().collect::<Vec<_>>())).is_err() {
+                                        warn!("Callback dropped for topic query {}. Results dropped", topic_hash);
+                                    }
+                                }
+                            }
+                        }
+                        TopicQueryState::Unsatisfied(topic_hash) => self.send_topic_queries(topic_hash, None),
+                    }
+                }
+                _ = registration_interval.tick() => {
+                    trace!("New registration interval, {}/{} topics to publish", topics_to_reg_iter.clone().count(), self.registration_attempts.len());
+                    let mut sent_regtopics = 0;
+                    let mut topic_item = topics_to_reg_iter.next();
+                    while let Some((topic, _topic_hash)) = topic_item {
+                        trace!("Publishing topic {} with hash {}", topic.topic(), topic.hash());
+                        sent_regtopics += self.send_register_topics(topic.clone());
+                        if sent_regtopics >= MAX_REGTOPICS_REGISTER_INTERVAL {
+                            break
+                        }
+                        topic_item = topics_to_reg_iter.next();
+                    }
+                    if topics_to_reg_iter.next().is_none() {
+                        topics_to_reg_iter = self.registration_attempts.keys().map(|topic| (topic.clone(), topic.hash())).collect::<Vec<(Topic, TopicHash)>>().into_iter();
+                    }
```

Comment:
done this way, removing a topic will have no effect over this until the rest of the topics are exhausted in the iterator

Reply:
Hm ok, this is my best attempt to get all topics registered. What happens if I try to send registrations for all kbuckets for all topics in one go is that Service is blocked for way too long from doing non-topic work and crashes. Any ideas on how to do this differently? Plugging the leak would be to check if a topic is removed when it is its turn in the frozen-iterator, i.e. if it is still relevant to register it, before making the registration attempts -maybe it's the best way?